File: lib/domain/models/library_status.dart
import 'package:libcheck/domain/models/availability_status.dart';

class LibraryStatus {
  const LibraryStatus({
    required this.systemId,
    required this.status,
    this.reserveUrl,
    required this.libKeyStatuses,
  });

  final String systemId;
  final AvailabilityStatus status;
  final String? reserveUrl;
  final Map<String, String> libKeyStatuses;

  AvailabilityStatus statusForLibKey(String libKey) {
    final apiString = libKeyStatuses[libKey];
    if (apiString == null) return AvailabilityStatus.notFound;
    return AvailabilityStatus.fromApiString(apiString);
  }
}

File: src/domain/models/libraryStatus.ts
import {
  AvailabilityStatus,
  availabilityFromApiString,
} from '@/domain/models/availabilityStatus';

export interface LibraryStatus {
  systemId: string;
  status: AvailabilityStatus;
  reserveUrl?: string;
  libKeyStatuses: Record<string, string>;
}

export function statusForLibKey(
  ls: LibraryStatus,
  libKey: string,
): AvailabilityStatus {
  const apiString = ls.libKeyStatuses[libKey];
  if (apiString === undefined) {
    // システム検索がエラーのとき libkey は空になる。蔵書なし(notFound)では
    // なく、システム全体の error 状態をそのまま分館にも反映する。
    return ls.status === AvailabilityStatus.error
      ? AvailabilityStatus.error
      : AvailabilityStatus.notFound;
  }
  return availabilityFromApiString(apiString);
}
